rush-plugins/rush-migrate-subspace-plugin/src/cleanSubspace.ts

```diff
@@ -1,14 +1,243 @@
+import { RESERVED_VERSIONS } from './constants/versions';
+
+const removeSupersetDependency = async (
+  subspaceName: string,
+  dependencyName: string,
+  versionsMap: Map<string, string[]>,
+  rootPath: string = getRootPath()
+): Promise<number> => {
+  const versions: string[] = Array.from(versionsMap.keys());
+  const subspaceCommonVersionsPath: string = getRushSubspaceCommonVersionsFilePath(subspaceName, rootPath);
+  const subspaceCommonVersionsJson: RushSubspaceCommonVersionsJson = loadRushSubspaceCommonVersions(
+    subspaceName,
+    rootPath
+  );
+
+  const newValidVersions: string[] = rSortVersions(versions).reduce<string[]>((prevVersions, currVersion) => {
+    const newVersions: string[] = [...prevVersions];
+    if (newVersions.includes(currVersion)) {
+      return newVersions;
+    }
+
+    const newSubsetVersion: string | undefined = newVersions.find((newVersion) =>
+      subsetVersion(newVersion, currVersion)
+    );
+
+    if (RESERVED_VERSIONS.includes(currVersion) || !newSubsetVersion) {
+      newVersions.push(currVersion);
+    } else {
+      // Update projects with new subset version
+      versionsMap.get(currVersion)?.forEach((projectName) => {
+        if (updateProjectDependency(projectName, dependencyName, newSubsetVersion, rootPath)) {
+          Console.debug(
+            `Updated project ${Colorize.bold(projectName)} for dependency ${Colorize.bold(
+              dependencyName
+            )} ${Colorize.bold(currVersion)} => ${Colorize.bold(newSubsetVersion)}!`
+          );
+        }
+      });
+    }
+
+    return newVersions;
+  }, []);
+
+  const removedAlternativeVersionsCount: number = versions.length - newValidVersions.length;
+  if (removedAlternativeVersionsCount > 0) {
+    // Update subspace common versions
+    if (newValidVersions.length > 0) {
+      subspaceCommonVersionsJson.allowedAlternativeVersions![dependencyName] = newValidVersions;
+    } else {
+      delete subspaceCommonVersionsJson.allowedAlternativeVersions![dependencyName];
+    }
+
+    JsonFile.save(subspaceCommonVersionsJson, subspaceCommonVersionsPath);
+  }
+
+  return removedAlternativeVersionsCount;
+};
+
+const removeDuplicatedDependencies = (subspaceName: string, rootPath: string = getRootPath()): void => {
+  Console.log(`Removing duplicated dependencies for subspace ${Colorize.bold(subspaceName)}...`);
+
+  const projects: IRushConfigurationProjectJson[] = queryProjectsFromSubspace(subspaceName, rootPath);
+  let countRemoved: number = 0;
+  projects.forEach((project) => {
+    const projectPackageFilePath: string = getProjectPackageFilePath(project.projectFolder, rootPath);
+    const projectPackageJson: IPackageJson = loadProjectPackageJson(project.projectFolder, rootPath);
+
+    const dependencies: string[] = Object.keys(projectPackageJson.dependencies || {});
+    const devDependencies: string[] = Object.keys(projectPackageJson.devDependencies || {});
+
+    devDependencies.forEach((devDependency) => {
```

Comment:
Everywhere, use `for (devDependency of devDependencies) {` instead of `forEach()`.  `forEach()` introduces a closure which is slower and error-prone.